### ISSUE TITLE
Kitchen sink vs. Scaffold approaches

### ISSUE THREAD

**Issue by sergi** [nearform/polaris#60]
Polaris is meant to be a development accelerator. That is, whenever you need to create a new mobile app, or a multi-platform web & mobile app, it should be our first choice (and hopefully the choice of many users, given its Open Source license).

At this point of the project, we need to decide between two main different approaches (and a bonus third approach!):

### Kitchen sink approach
In this approach we'd provide an example app with an opinionated choice of libraries and approaches, and a lot of screens and components that demonstrate different components, pre-made functionality, etc. For example, we could have a screen that shows how to capture QR codes and do a POST request to a fake server, or a screen with a dnd list, offline functionality activated, etc.

Pros:
- Great for prototype development and PoCs
- Great for non-expert React Native developers

Cons:
- More complex, heavy app
- You'd have to spend time discarding the bits you won't use
- Might get in the way of expert developers
- More prone to obsolescence
- More maintenance

### Scaffold approach
In this approach we'd provide a skeleton app with the basic dev workflow set-up, and some conventions and configurations already in place. The app itself would include our choice of functionality and libraries (e.g. authentication), but it would be sparse. It would have stuff like CI/CD, E2E, etc. already in place though, where developers would only need to change a couple of parameters to make it work for their environment.

Pros:
- Maximum flexibility
- Provides opinionated, minimal common baseline for projects
- Great for experienced developers
- Minimal maintenance

Cons:
- Not that much advantage over something like Expo
- Some common tasks would still have to be build by developers
- Not that useful for inexperienced developers

### Extensible approach
In this approach we'd provide a scaffold approach that can be extended with plugins, which can be hosted in external repos.

Pros:
- Best of both worlds?
- Decentralizes development and takes pressure off potential OS project maintenance

Cons:
- We need to architect a plugin system that works
- The assumption that people would develop plugins for the system might be optimistic

What are your thoughts?

**Comment by skywickenden** [comment]
Would a middle road be possible: Take a scaffold approach, but include a docs section to detail what would be in the kitchen sink  - with code snippets for rapidly getting up to speed. 

Pros:
- Pros being able to rapidly get started
- Inexperienced developers able to get clear guidance on good practice
- Teaches inexperienced devs why as well as how
- Kitchen sink variations can be detailed
- Less maintenance to keep the code up to date
- Forking just requires deleting the docs folder rather than taking features out.

Cons:
- Would require effort to keep the docs up to date
- Good docs are hard to write

**Comment by AlanSl** [comment]
I agree with middle ground in principle, but the trouble with doing it by docs is they become out of date quickly as new work is done on the core, and it's very easy to to write docs that sound good in theory but don't really work in practice.

What I'd really like is some kind of monorepo structure like:

```
/polaris
  /src
     /* core stuff here */
  /examples
     /camera-integration
        /* starts expo using core, adds camera integration views */
     /complex-navigation
        /* starts expo using core, adds layout with lots of nested navigators */
     /advanced-theme
        /* starts expo using core, adds theme that shows advanced features */
     /more-examples...
```

etc etc

Then this would keep itself update through tests, and devs can copy and paste the bits they want to use. It has most of the benefits of the pluggable option without needing to design and maintain a plugin system.

The question is how to do this with Expo... I *think* it could be done using something like yarn workspaces with each example being a small workspace with a few components, routes and a webpack config that pulls in the main webpack config and adds a few aliases for its own additions.

**Comment by skywickenden** [comment]
@AlanSl If we can make that work it gets my vote. 

There would be a danger of a bunch of half baked examples. They would also need reviewing regularly - the tests might pass but best practice might have moved on.

The two advantages I can think of for the docs approach is that it would take less work and more easily allow for conflicting approaches to be documented.

**Comment by lpww** [comment]
I don't have the full context but will throw a curveball as I think scaffold could be the way to go. As far as I understand, the main goal for Polaris is to accelerate internal projects. This means the target user base is going to be more experienced developers who just need the basics to get started. It doesn't need to be vastly different to Expo to achieve our primary goal.

If we are talking about it's position in the open source community, then I agree the combined approach is a better value proposition. However, I don't think that's the main objective here. We could always start with a scaffold and build out examples as we need them later

**Comment by fiacc** [comment]
I like the extensible approach. A solid scaffold to get started and examples of functionality will definitely help on putting together MVPs. That said any examples should provide more value than just what you get if you google it. An opinionated plugin architecture is also something that would be valuable, it tends to come up in projects.

**Comment by mcollina** [comment]
I like @AlanSl approach and structure in https://github.com/nearform/polaris/issues/60#issuecomment-660125586. 

This is probably similar to the "extensible" approach in some form.
